### ISSUE TITLE
Deadlocks causing duplicate payments

### ISSUE THREAD

**Issue by andrew-smart** [mailchimp/mc-magento#310]
An issue has come to light whereby customers are getting charged twice but only a single order appearing in Magento.

Looking through the exception logs, all of the issues reported have been caused by deadlocking in the database and the order never reaching the database. Generally customers are then getting thrown back to the cart and trying again.

This is the message we are getting;

```sql
Next Zend_Db_Statement_Exception: SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction, query was: UPDATE `mailchimp_ecommerce_sync_data` SET `related_id` = ?, `type` = ?, `mailchimp_store_id` = ?, `mailchimp_sync_error` = ?, `mailchimp_sync_delta` = '2017-06-02 13:50:05', `mailchimp_sync_modified` = ?, `mailchimp_sync_deleted` = ?, `mailchimp_token` = ?, `batch_id` = ? WHERE (id='55521') in /home/mymemory1/public_html/lib/Zend/Db/Statement/Pdo.php:235
```

This is happening during the sales_order_place_after observer call (so just after the payment has been taken).

Potentially a sync of the data is happening during the placement of a new order.

**Comment by centerax** [comment]
what version are you on?

**Comment by andrew-smart** [comment]
1.1.7

**Comment by centerax** [comment]
I think the deadlocks are related to something else because the hook to [sales_order_place_before](https://github.com/mailchimp/mc-magento/blob/develop/app/code/community/Ebizmarts/MailChimp/etc/config.xml#L194-L201) is not pushing nothing to MailChimp.

**Comment by andrew-smart** [comment]
The issue is triggered in the sales_order_place_**after** event when it calls saveEcommerceSyncData.

**Comment by centerax** [comment]
Yes, thats not correct.

We will probably move this to a cronjob.

**Comment by centerax** [comment]
@andrew-smart can you send us the full stack trace to mailchimp@ebizmarts-desk.zendesk.com ?

Thanks.

**Comment by Santiagoebizmarts** [comment]
@andrew-smart If you could also let us know if you have a custom checkout and if you do which one would be great!

Regards.

**Comment by andrew-smart** [comment]
I have sent the data.

**Comment by elluminatte** [comment]
@centerax @Santiagoebizmarts hello guys. Faced the same problem. I've sent data to mailchimp@ebizmarts-desk.zendesk.com
To my mind this is critical issue, customers potentially lose their money.
Seems you know about it and do not fix anything.. Hope you'll work with this soon

Regards, Nikolay

**Comment by Santiagoebizmarts** [comment]
Hello @andrew-smart 

This has been fixed and the solution will be available in the next version 1.1.8 which will be released soon.

You can see the fix for this [here](https://github.com/mailchimp/mc-magento/commit/4856872733b3b32b508a8d03da480e7506a13124).

Regards.
